File: internal/cache/cache.go
package cache

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

type Pack struct {
	Dir  string
	Name string
}

func FindPremiumCache(minecraftDir string) (string, error) {
	if minecraftDir != "" {
		cache := filepath.Join(minecraftDir, "premium_cache", "skin_packs")
		if info, err := os.Stat(cache); err == nil && info.IsDir() {
			return cache, nil
		}
		return "", fmt.Errorf("not found: %s", cache)
	}

	appdata := os.Getenv("APPDATA")
	if appdata != "" {
		cache := filepath.Join(appdata, "Minecraft Bedrock", "premium_cache", "skin_packs")
		if info, err := os.Stat(cache); err == nil && info.IsDir() {
			return cache, nil
		}
	}

	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("could not find premium_cache directory")
	}
	wineBase := filepath.Join(home, ".local", "share", "bedrock-on-linux", "compatdata", "pfx", "drive_c", "users")
	matches, _ := filepath.Glob(filepath.Join(wineBase, "*"))
	for _, userDir := range matches {
		cache := filepath.Join(userDir, "AppData", "Roaming", "Minecraft Bedrock", "premium_cache", "skin_packs")
		if info, err := os.Stat(cache); err == nil && info.IsDir() {
			return cache, nil
		}
	}

	return "", fmt.Errorf("could not find premium_cache directory")
}

func ListPacks(cacheDir string) ([]Pack, error) {
	entries, err := os.ReadDir(cacheDir)
	if err != nil {
		return nil, err
	}

	var packs []Pack
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		if strings.HasSuffix(entry.Name(), ".backup") {
			continue
		}
		packDir := filepath.Join(cacheDir, entry.Name())
		manifestPath := filepath.Join(packDir, "manifest.json")
		if _, err := os.Stat(manifestPath); os.IsNotExist(err) {
			continue
		}

		name := entry.Name()
		data, err := os.ReadFile(manifestPath)
		if err == nil {
			var manifest struct {
				Header struct {
					Name string `json:"name"`
				} `json:"header"`
			}
			if json.Unmarshal(data, &manifest) == nil && manifest.Header.Name != "" {
				name = manifest.Header.Name
			}
		}
		packs = append(packs, Pack{Dir: packDir, Name: name})
	}

	sort.Slice(packs, func(i, j int) bool {
		return strings.ToLower(packs[i].Name) < strings.ToLower(packs[j].Name)
	})
	return packs, nil
}

File: skinx/cache.py
"""Minecraft Bedrock premium_cache discovery and pack listing."""

import os
from pathlib import Path


def find_premium_cache(minecraft_dir=None) -> Path:
    """Locate the premium_cache/skin_packs directory.

    Searches in order:
    1. *minecraft_dir* if provided (``<dir>/premium_cache/skin_packs``)
    2. Native Windows (%APPDATA%)
    3. BedrockOnLinux Wine prefix (https://github.com/Wyze3306/BedrockOnLinux)
    """
    if minecraft_dir:
        cache = Path(minecraft_dir) / "premium_cache/skin_packs"
        if cache.exists():
            return cache
        raise FileNotFoundError(f"Not found: {cache}")

    appdata = os.environ.get("APPDATA")
    if appdata:
        cache = Path(appdata) / "Minecraft Bedrock/premium_cache/skin_packs"
        if cache.exists():
            return cache
    bedrock_wine = Path.home() / ".local/share/bedrock-on-linux/compatdata/pfx/drive_c/users"
    for user_dir in bedrock_wine.glob("*"):
        cache = user_dir / "AppData/Roaming/Minecraft Bedrock/premium_cache/skin_packs"
        if cache.exists():
            return cache
    raise FileNotFoundError("Could not find premium_cache directory")


def list_packs(cache: Path):
    """List all valid skin packs in the cache directory.

    Returns list of (pack_dir, display_name) sorted alphabetically.
    """
    packs = []
    for pack_dir in cache.iterdir():
        if pack_dir.is_dir() and (pack_dir / "manifest.json").exists() and not pack_dir.name.endswith(".backup"):
            try:
                with open(pack_dir / "manifest.json") as f:
                    import json
                    manifest = json.load(f)
                name = manifest.get("header", {}).get("name", pack_dir.name)
            except Exception:
                name = pack_dir.name
            packs.append((pack_dir, name))
    packs.sort(key=lambda x: x[1].lower())
    return packs
